Update SimpleChannel to make StreamCodecs easier

LexManos opened the pull request:
Split into a couple different commits, the first is the general cleanup allowing correctly types FriendlyByteBuf or RegistryFriendlyButeBuf classes to be used in encode and decode functions.

Second adds the codec(StreamCodec) function to MessageBuilder, which is shorthand for setting the encoder and decoder.
As well as fix the protocols related to Forge's packets.
And add a build() function to SimpleChannel, which prevents additional packets from being registered. Like a normal builder should...

Third expands the SimpleChannel builders to make things more streamlined, some examples of the new builders

```java

    CHANNEL = ChannelBuilder
        .named(NAME)
        .optional()
        .networkProtocolVersion(0)
        .simpleChannel()
            .handler(CONTEXT)
                .login()
                    .serverbound()
                        .add(LoginWrapper.class, LoginWrapper.STREAM_CODEC)
                    
        .buildAll();
 ```

```java

    CHANNEL = ChannelBuilder
        .named(NAME)
        .optional()
        .networkProtocolVersion(0)
        .simpleChannel()
            .handler(CONTEXT)
                .login()
                    .serverbound()
                        .add(ServerToClientLogin.class, ServerToClientLogin.STREAM_CODEC)
                    .build()
                    .clientbound()
                        .add(ClientToServerLogin.class, ClientToLoginServer.STREAM_CODEC)
                    .build()
                .build()
                .play()
                    .add(BiDirectionPlay.class, BiDirectionPlay.STREAM_CODEC)
                .build()
            .build()
            .config()
                .add(BiDirectionConfigWithoutContext.class, BiDirectionConfigWithoutContext.STREAM_CODEC, BidrectionConfig::handle)
        .buildAll();
```
        
        
The [4th one](https://github.com/MinecraftForge/MinecraftForge/pull/9959/commits/bc665eb2e8bce88a8b9e702267d80654206c41a6) codecify's all of Forge's packets.  Honestly i like it but I also don't. But it shows how simple it is.

LexManos committed:
Rework networking so that RegistryFriendlyByteBuf is useable for modders

LexManos committed:
Simple support for StreamCodec in SimpleChannel

LexManos committed:
Add more Simple Channel builders to make using codecs easier.

LexManos committed:
Codecify all Forge packets

MrCrayfish commented on src/main/java/net/minecraftforge/network/NetworkDirection.java:
```diff
@@ -5,88 +5,20 @@
+    public static final NetworkDirection<FriendlyByteBuf> LOGIN_TO_CLIENT = new NetworkDirection<>(NetworkProtocol.LOGIN, PacketFlow.CLIENTBOUND);
+    public static final NetworkDirection<FriendlyByteBuf> LOGIN_TO_SERVER = new NetworkDirection<>(NetworkProtocol.LOGIN, PacketFlow.SERVERBOUND);
+    public static final NetworkDirection<FriendlyByteBuf> CONFIG_TO_CLIENT = new NetworkDirection<>(NetworkProtocol.CONFIG, PacketFlow.CLIENTBOUND);
+    public static final NetworkDirection<FriendlyByteBuf> CONFIG_TO_SERVER = new NetworkDirection<>(NetworkProtocol.CONFIG, PacketFlow.SERVERBOUND);
```
Sort of out of scope of this PR but just general feedback.

In terms of networking, "config" should be expressed as its full name "configuration". It would help with readability and usability since you would not be switching between the two. This would apply everywhere in the networking code, not just these lines. Forge is also already using the fullname for the following `GatherLoginConfigurationTasksEvent`, `ConfigurationTaskContext`, and `SimpleConfigurationTask`.

MrCrayfish commented on src/main/java/net/minecraftforge/network/SimpleChannel.java:
```diff
@@ -271,26 +407,326 @@ public SimpleChannel add() {
+            checkBuilt();
+            this.built = true;
+            return parent;
+        }
```
Readability wise, I think the chaining starts to go to deep. I think simplification would greatly improve this and just having a single `build` call at the end.

```java
CHANNEL = ChannelBuilder
    .named(NAME)
    .optional()
    .networkProtocolVersion(0)
    .simpleChannel()
    .login(CONTEXT) // Messages here must follow adhere to context
        .serverbound(ServerToClientLogin.class, ServerToClientLogin.STREAM_CODEC)
        .clientbound(ClientToServerLogin.class, ClientToLoginServer.STREAM_CODEC)
        .add(ClientToServerLogin.class, ClientToLoginServer.STREAM_CODEC, NetworkDirection.LOGIN_TO_SERVER) // Explicit version
    .login() // Implicitly calls protocal#build and starts a new protocol?
        .clientbound(ClientExample.class, ClientExample.STREAM_CODEC, ClientExample::handle)    
    .play(CONTEXT)
        .bidirectional(BiDirectionPlay.class, BiDirectionPlay.STREAM_CODEC) // Named "bidirectional" instead of "add"
    .configuration()
        .bidirectional(BiDirectionConfigWithoutContext.class, BiDirectionConfigWithoutContext.STREAM_CODEC, BidrectionConfig::handle)
    .build();
```
    
I realized the above might create a nightmare of a class, so using wrapping msg register under a function might be easier. However I do still think the above is cleaner and more preferable.
```java
CHANNEL = ChannelBuilder
    .named(NAME)
    .optional()
    .networkProtocolVersion(0)
    .simpleChannel()
    .login(CONTEXT, protocol -> {
        protocol.serverbound(ServerToClientLogin.class, ServerToClientLogin.STREAM_CODEC);
        protocol.clientbound(ClientToServerLogin.class, ClientToLoginServer.STREAM_CODEC);
        protocol.add(ClientToServerLogin.class, ClientToLoginServer.STREAM_CODEC, NetworkDirection.LOGIN_TO_SERVER); 
    })
    .login(protocol -> {
        protocol.clientbound(ClientExample.class, ClientExample.STREAM_CODEC, ClientExample::handle);    
    })
    .play(CONTEXT, protocol -> {
        protocol.bidirectional(BiDirectionPlay.class, BiDirectionPlay.STREAM_CODEC);  
    })
    .configuration(protocol -> {
        protocol.bidirectional(BiDirectionConfigWithoutContext.class, BiDirectionConfigWithoutContext.STREAM_CODEC, BidrectionConfig::handle);    
    })
    .build();```

LexManos committed:
Move builder api to their own files to prevent SimpleChannel from getting even larger.
Add support for consumer based builders.
Allowed flow/protocol be specified in any order.

LexManos committed:
Revert to explicitly seperate states. Meaning you have to call build to go from flow->protocol or protocol->connection.
The cleaner code {removal of 1 .build() call} wasn't worth the potential screwups and ugly code that it produced

LexManos commented:
Not arguing the config change, I dont care either way. so changed it to the full `configuration`

Not going to go with the serverbound/clientbound/bidirectional methods because it would mean that all add functions need to be duplicated for all flows. And that is just asking for code bloat {see the old MessageBuilder class}.

They are currently duplicated in Connection for bidirectional packets, but im not convinced that's the best idea. It just saves a `.flow(null)` call, which i could add as a `.unbound()` or something. Doing so would save maintenance cost.

I think this is gunna be the final design until I rip out MessageBuilder. In theory tho that old API is untouched/binary compatible {haven't tested} at the very least it's source compatible.

I think that the examples provided are clean enough. This being Forge's handler:
```java
    public static SimpleChannel CONFIG = ChannelBuilder
        .named(HANDSHAKE_NAME)
        .optional()
        .networkProtocolVersion(0)
        .simpleChannel()

            .configuration()
                .add(ModVersions.class, ModVersions.STREAM_CODEC, ctx(ForgePacketHandler::handleModVersions))
                .add(ChannelVersions.class, ChannelVersions.STREAM_CODEC, ctx(ForgePacketHandler::handleChannelVersions))
                .serverbound()
                    .add(Acknowledge.class, Acknowledge.STREAM_CODEC, ctx(ForgePacketHandler::handleClientAck))
                .build()
                .clientbound()
                    .add(RegistryList.class, RegistryList.STREAM_CODEC, ctx(ForgePacketHandler::handleRegistryList))
                    .add(RegistryData.class, RegistryData.STREAM_CODEC, ctx(ForgePacketHandler::handleRegistryData))
                    .add(ConfigData.class, ConfigData.STREAM_CODEC, ctx(ForgePacketHandler::handleConfigSync))
                    .add(MismatchData.class, MismatchData.STREAM_CODEC, ctx(ForgePacketHandler::handleModMismatchData))

        .buildAll();     
```

MrCrayfish committed:
Further clean up of networking code

MrCrayfish committed:
Renamed SimpleHandlerProtocol

MrCrayfish committed:
Move protocol methods to a single interface

MrCrayfish committed:
Add licensing and remove author comments

MrCrayfish committed:
Removed unused generic

MrCrayfish committed:
Change self method to cast with unchecked warning

MrCrayfish committed:
Remove unnecessary implement on simplechannel

LexManos committed:
Move Handler interface into their own package to better illustrate/group together.
Remove IFlow marker
Merge SimpleProtocolFactory into SimpleConnection

MrCrayfish committed:
Fix formatting

MrCrayfish committed:
Remove private docs.

It's already in the super interface

MrCrayfish committed:
Rename IFlow/IProtocol to BaseFlow/BaseProtocol

MrCrayfish committed:
Revert "Rename IFlow/IProtocol to BaseFlow/BaseProtocol"

This reverts commit b0b1f8eba1b46583f3aaa7d8bf646803e607fdac.

MrCrayfish committed:
Rename IFlow/IProtocol to BaseFlow/BaseProtocol

MrCrayfish committed:
Use fullname for generic argument

LexManos committed:
Bit of code cleanup, make SimpleContext not care about Handler Key and remove need for private static methods

LexManos committed:
Add 'addMain' to SimpleFlow

LexManos committed:
Add flow(Consumer) to BaseProtocol, as an alias for bidirectional(Consumer)

LexManos committed:
Javadocs cleanup and add consumer versions of the context aware protocol bouncers

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=9959) <br/>All committers have signed the CLA.

LexManos committed:
Fix order in which Forge's packets are registered making sure they get the same discriminators as before this rework.
Properly mark OpenContainers packet as S->C only

LexManos committed:
Add clarification and warning javadocs to SimpleBuildable

LexManos commented:
Alright, updated with MrCrayfish's suggestions.
It simplifies the code quite a bit not needing to pass around the parent object. At the cost of only calling build() once. 
Which isn't a big deal, with the way the system works it's basically just making disjointed builders with slightly modified contexts. 

Anyways, as soo as @MrCrayfish signs the CLA This can be pulled

PaintNinja commented on src/main/java/net/minecraftforge/network/packets/MismatchData.java:
```diff
@@ -22,6 +23,7 @@ public record MismatchData(
     Map<ResourceLocation, Version> mismatched,
     Set<ResourceLocation> missing
 ) {
+    public static StreamCodec<FriendlyByteBuf, MismatchData> STREAM_CODEC = StreamCodec.ofMember(MismatchData::encode, MismatchData::decode);
```
```suggestion
    public static final StreamCodec<FriendlyByteBuf, MismatchData> STREAM_CODEC = StreamCodec.ofMember(MismatchData::encode, MismatchData::decode);
```
Missing final

PaintNinja commented on src/main/java/net/minecraftforge/network/packets/ModVersions.java:
```diff
@@ -16,6 +17,8 @@
  * The mod data is stored as follows: [modId -> [modName, modVersion]]
  */
 public record ModVersions(Map<String, Info> mods) {
+    public static StreamCodec<FriendlyByteBuf, ModVersions> STREAM_CODEC = StreamCodec.ofMember(ModVersions::encode, ModVersions::decode);
```
```suggestion
    public static final StreamCodec<FriendlyByteBuf, ModVersions> STREAM_CODEC = StreamCodec.ofMember(ModVersions::encode, ModVersions::decode);
```
Missing final

PaintNinja commented on src/main/java/net/minecraftforge/network/packets/OpenContainer.java:
```diff
@@ -12,22 +12,24 @@
 
 // TODO: Reevaluate if this is needed This is the same as ClientboundOpenScreenPacket packet but allows for additional data
 public class OpenContainer {
+    public static StreamCodec<RegistryFriendlyByteBuf, OpenContainer> STREAM_CODEC = StreamCodec.ofMember(OpenContainer::encode, OpenContainer::decode);
```
```suggestion
    public static final StreamCodec<RegistryFriendlyByteBuf, OpenContainer> STREAM_CODEC = StreamCodec.ofMember(OpenContainer::encode, OpenContainer::decode);
```
Missing final

PaintNinja commented on src/main/java/net/minecraftforge/network/packets/RegistryData.java:
```diff
@@ -6,10 +6,13 @@
 import net.minecraftforge.registries.ForgeRegistry.Snapshot;
 
 public record RegistryData(int token, ResourceLocation name, Snapshot data) {
+    public static StreamCodec<FriendlyByteBuf, RegistryData> STREAM_CODEC = StreamCodec.ofMember(RegistryData::encode, RegistryData::decode);
```
```suggestion
    public static final StreamCodec<FriendlyByteBuf, RegistryData> STREAM_CODEC = StreamCodec.ofMember(RegistryData::encode, RegistryData::decode);
```
Missing final

PaintNinja commented on src/main/java/net/minecraftforge/network/packets/RegistryList.java:
```diff
@@ -18,6 +19,8 @@ public record RegistryList(
     List<ResourceLocation> normal,
     List<ResourceKey<? extends Registry<?>>> datapacks) {
 
+    public static StreamCodec<FriendlyByteBuf, RegistryList> STREAM_CODEC = StreamCodec.ofMember(RegistryList::encode, RegistryList::decode);
```
```suggestion
    public static final StreamCodec<FriendlyByteBuf, RegistryList> STREAM_CODEC = StreamCodec.ofMember(RegistryList::encode, RegistryList::decode);
```
Missing final

PaintNinja commented on src/main/java/net/minecraftforge/network/packets/SpawnEntity.java:
```diff
@@ -34,6 +34,7 @@
  */
 // TODO: Re-write this packet into something simpler. This is literally just ClientboundAddEntityPacket with an extra byte[]
 public class SpawnEntity {
+    public static StreamCodec<RegistryFriendlyByteBuf, SpawnEntity> STREAM_CODEC = StreamCodec.ofMember(SpawnEntity::encode, SpawnEntity::decode);
```
```suggestion
    public static final StreamCodec<RegistryFriendlyByteBuf, SpawnEntity> STREAM_CODEC = StreamCodec.ofMember(SpawnEntity::encode, SpawnEntity::decode);
```
Missing final

LexManos committed:
Update src/main/java/net/minecraftforge/network/packets/MismatchData.java

Co-authored-by: Paint_Ninja <PaintNinja@users.noreply.github.com>

LexManos committed:
Update src/main/java/net/minecraftforge/network/packets/ModVersions.java

Co-authored-by: Paint_Ninja <PaintNinja@users.noreply.github.com>

LexManos committed:
Update src/main/java/net/minecraftforge/network/packets/OpenContainer.java

Co-authored-by: Paint_Ninja <PaintNinja@users.noreply.github.com>

LexManos committed:
Update src/main/java/net/minecraftforge/network/packets/RegistryData.java

Co-authored-by: Paint_Ninja <PaintNinja@users.noreply.github.com>

LexManos committed:
Update src/main/java/net/minecraftforge/network/packets/RegistryList.java

Co-authored-by: Paint_Ninja <PaintNinja@users.noreply.github.com>

LexManos committed:
Update src/main/java/net/minecraftforge/network/packets/SpawnEntity.java

Co-authored-by: Paint_Ninja <PaintNinja@users.noreply.github.com>

PaintNinja commented on src/main/java/net/minecraftforge/network/NetworkInitialization.java:
```diff
@@ -5,113 +5,79 @@
+    private static final ResourceLocation PLAY_NAME = new ResourceLocation("forge", "network");
     public static final AttributeKey<ForgePacketHandler> CONTEXT = AttributeKey.newInstance(HANDSHAKE_NAME.toString());
 
     public static SimpleChannel LOGIN = ChannelBuilder
```
```suggestion
    public static final SimpleChannel LOGIN = ChannelBuilder
```
I know this is inside a class marked as internal, but would be wise to mark this as final to prevent modders from screwing with it anyway. Plus muh static finals perf

PaintNinja commented on src/main/java/net/minecraftforge/network/NetworkInitialization.java:
```diff
@@ -5,113 +5,79 @@
-            .add();
-
-    public static SimpleChannel PLAY = ChannelBuilder
+    public static SimpleChannel CONFIG = ChannelBuilder
```
```suggestion
    public static final SimpleChannel CONFIG = ChannelBuilder
```

PaintNinja commented on src/main/java/net/minecraftforge/network/NetworkInitialization.java:
```diff
@@ -5,113 +5,79 @@
+                    .add(OpenContainer.class, OpenContainer.STREAM_CODEC, OpenContainer::handle)
+        .build();
+
+    public static SimpleChannel PLAY = CONFIG;;
```
```suggestion
    public static final SimpleChannel PLAY = CONFIG;;
```

PaintNinja commented on src/main/java/net/minecraftforge/network/NetworkInitialization.java:
```diff
@@ -5,113 +5,79 @@
+import java.util.function.BiConsumer;
 
 @ApiStatus.Internal
 public class NetworkInitialization {
```
```suggestion
public final class NetworkInitialization {
```
Internal class that we don't extend, so mark as final to prevent unsupported extending of internals.

PaintNinja commented on src/main/java/net/minecraftforge/network/NetworkInstance.java:
```diff
@@ -28,7 +28,9 @@
  */
 @ApiStatus.Internal
 public class NetworkInstance {
```
```suggestion
public final class NetworkInstance {
```
Internal class we don't extend ourselves, to make final to prevent unsupported extending of it

PaintNinja commented on src/main/java/net/minecraftforge/network/packets/ConfigData.java:
```diff
@@ -6,8 +6,11 @@
+import net.minecraft.network.codec.StreamCodec;
 
 public record ConfigData(String name, byte[] data) {
+    public static StreamCodec<FriendlyByteBuf, ConfigData> STREAM_CODEC = StreamCodec.ofMember(ConfigData::encode, ConfigData::decode);
```
```suggestion
    public static final StreamCodec<FriendlyByteBuf, ConfigData> STREAM_CODEC = StreamCodec.ofMember(ConfigData::encode, ConfigData::decode);
```

LexManos committed:
Update src/main/java/net/minecraftforge/network/NetworkInitialization.java

Co-authored-by: Paint_Ninja <PaintNinja@users.noreply.github.com>

LexManos committed:
Update src/main/java/net/minecraftforge/network/NetworkInitialization.java

Co-authored-by: Paint_Ninja <PaintNinja@users.noreply.github.com>

LexManos committed:
Update src/main/java/net/minecraftforge/network/NetworkInitialization.java

Co-authored-by: Paint_Ninja <PaintNinja@users.noreply.github.com>

LexManos committed:
Update src/main/java/net/minecraftforge/network/NetworkInstance.java

Co-authored-by: Paint_Ninja <PaintNinja@users.noreply.github.com>

LexManos committed:
Update src/main/java/net/minecraftforge/network/packets/ConfigData.java

Co-authored-by: Paint_Ninja <PaintNinja@users.noreply.github.com>

LexManos committed:
Update src/main/java/net/minecraftforge/network/NetworkInitialization.java

Co-authored-by: Paint_Ninja <PaintNinja@users.noreply.github.com>

MrCrayfish commented:
CLA signed. I've gone over the additional changes and everything looks good. There are some broken `@links` in the documentation of `SimpleChannel$MessageBuilder` and `SimpleConnection` that maybe you want to address for this PR.

LexManos committed:
Fix a couple javadoc links

LexManos committed:
More Javadoc stuff
